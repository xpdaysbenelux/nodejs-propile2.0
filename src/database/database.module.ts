import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { UserRepository, RoleRepository } from './repositories';
import { User, Role } from './entities';
import { Config } from '../config';

@Module({
    imports: [
        TypeOrmModule.forRoot({
            ...Config.database,
            entities: [Role, User],
        }),
        TypeOrmModule.forFeature([RoleRepository, UserRepository]),
    ],
    exports: [TypeOrmModule],
})
export class DatabaseModule {}
