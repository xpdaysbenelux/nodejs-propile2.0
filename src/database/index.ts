export * from './entities';
export * from './repositories';
export * from './database.module';
