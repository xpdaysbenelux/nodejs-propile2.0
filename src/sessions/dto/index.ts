export * from './create-session.dto';
