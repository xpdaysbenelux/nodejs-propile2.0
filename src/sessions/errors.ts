import { BadRequestException } from '@nestjs/common';

export class SessionTitleAlreadyInUse extends BadRequestException {
    constructor() {
        super(
            'A session with this title already exists',
            'SESSION_TITLE_ALREADY_IN_USE',
        );
    }
}
