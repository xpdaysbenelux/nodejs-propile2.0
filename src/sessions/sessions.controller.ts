import {
    Controller,
    Post,
    Body,
    Param,
    UseGuards,
    Get,
    Put,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { SessionsService } from './sessions.service';
import {
    CreateSessionRequest,
    UpdateSessionRequest,
    SessionIdParam,
    SessionResponse,
} from './dto';
import {
    Origin,
    RequiredPermissions,
    UserSession,
} from '../_shared/decorators/';
import { RequiredPermissionsGuard } from '../_shared/guards';
import { IUserSession } from '../_shared/constants';
import { SessionsQueries } from './sessions.queries';
import { UserIdParam } from '../users/dto';
import { GetSessionsResponse } from './dto/get-sessions.dto';

@Controller('sessions')
@ApiTags('sessions')
export class SessionsController {
    constructor(
        private readonly sessionService: SessionsService,
        private readonly sessionQueries: SessionsQueries,
    ) {}

    /**
     * Gets the sessions from a specific user.
     * @param userId
     */
    @RequiredPermissions({ sessions: { view: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get(':userId')
    getSessionsOfUser(
        @Param() params: UserIdParam,
    ): Promise<GetSessionsResponse> {
        return this.sessionQueries.getSessionsByUser(params.userId);
    }

    @RequiredPermissions({ sessions: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get(':sessionId')
    getSessionById(
        @Param() params: SessionIdParam,
        @UserSession() session: IUserSession,
    ): Promise<SessionResponse> {
        return this.sessionQueries.getSession(params.sessionId);
    }

    @Post()
    async createSession(
        @Body() body: CreateSessionRequest,
        @Origin() origin: string,
    ): Promise<void> {
        await this.sessionService.createSession(body, origin);
    }

    @RequiredPermissions({ sessions: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Put(':sessionId')
    async updateSession(
        @Body() body: UpdateSessionRequest,
        @Param() params: SessionIdParam,
        @UserSession() userSession: IUserSession,
    ): Promise<string> {
        const sessionId = await this.sessionService.updateSession(
            body,
            params.sessionId,
            userSession,
        );
        return sessionId;
    }
}
