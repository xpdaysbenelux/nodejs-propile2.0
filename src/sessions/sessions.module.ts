import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';

import { SessionsController } from './sessions.controller';
import { SessionsService } from './sessions.service';
import { DatabaseModule } from '../database';
import { Config } from '../config';
import { SessionsQueries } from './sessions.queries';
import { MailerModule } from '../mailer/mailer.module';

@Module({
    imports: [
        DatabaseModule,
        JwtModule.register({
            secret: Config.jwt.secret,
        }),
        MailerModule,
    ],
    controllers: [SessionsController],
    providers: [SessionsService, SessionsQueries],
})
export class SessionsModule {}
