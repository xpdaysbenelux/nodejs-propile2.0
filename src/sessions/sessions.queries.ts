import { Injectable } from '@nestjs/common';

import { SessionRepository, Session } from '../database';
import { SessionResponse } from './dto';
import { GetSessionsResponse } from './dto/get-sessions.dto';
import { SelectQueryBuilder } from 'typeorm';

const sessionFields = [
    'session.id',
    'session.createdAt',
    'session.title',
    'session.subTitle',
    'session.xpFactor',
    'session.description',
    'session.shortDescription',
    'session.sessionState',
    'session.goal',
    'session.type',
    'session.topic',
    'session.maxParticipants',
    'session.duration',
    'session.laptopsRequired',
    'session.otherLimitations',
    'session.roomSetup',
    'session.neededMaterials',
    'session.expierenceLevel',
    'session.outline',
    'session.materialDescription',
    'session.materialUrl',
    'firstPresenter.id',
    'firstPresenter.email',
    'secondPresenter.id',
    'secondPresenter.email',
];

@Injectable()
export class SessionsQueries {
    constructor(private readonly sessionRepository: SessionRepository) {}

    async getSession(sessionId: string): Promise<SessionResponse> {
        return this.sessionRepository
            .createQueryBuilder('session')
            .select(sessionFields)
            .where('session.id = :sessionId', { sessionId })
            .innerJoin('session.firstPresenter', 'firstPresenter')
            .leftJoin('session.secondPresenter', 'secondPresenter')
            .getOne();
        // TODO leftJoin voor intendedAudience
    }

    async getSessions(): Promise<GetSessionsResponse> {
        const [sessions, totalCount] = await this.selectSessionColumns(
            this.sessionRepository.createQueryBuilder('session'),
        ).getManyAndCount();

        return {
            meta: {
                count: sessions.length,
                totalCount,
                skip: null,
            },
            data: sessions,
        };
    }

    private selectSessionColumns(
        queryBuilder: SelectQueryBuilder<Session>,
    ): SelectQueryBuilder<Session> {
        return queryBuilder
            .select(sessionFields)
            .innerJoin('session.firstPresenter', 'firstPresenter')
            .leftJoin('session.secondPresenter', 'secondPresenter');
    }
}
