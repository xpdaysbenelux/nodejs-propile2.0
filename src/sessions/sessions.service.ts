import { Injectable } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';

import {
    UserRepository,
    RoleRepository,
    User,
    SessionRepository,
    Session,
    PersonaRepository,
} from '../database';
import { CreateSessionRequest } from './dto/create-session.dto';
import { MailerService } from '../mailer/mailer.service';
import {
    SessionTitleAlreadyInUse,
    SessionNotFound,
    SessionEditNotAllowed,
    SessionPresenterEmailsMustDiffer,
} from './errors';
import {
    registerMessage,
    sessionCreatedForFirstPresenterMessage,
    sessionCreatedForSecondPresenterMessage,
} from '../mailer/messages';
import { UpdateSessionRequest } from './dto';
import { IUserSession } from '../_shared/constants';
import { UserState } from '../_shared/constants';
import { In } from 'typeorm';

@Injectable()
export class SessionsService {
    constructor(
        private readonly jwtService: JwtService,
        private readonly mailerService: MailerService,
        private readonly userRepository: UserRepository,
        private readonly roleRepository: RoleRepository,
        private readonly sessionRepository: SessionRepository,
        private readonly personaRepository: PersonaRepository,
    ) {}

    async createSession(
        body: CreateSessionRequest,
        origin: string,
    ): Promise<void> {
        const {
            title,
            subTitle,
            emailFirstPresenter,
            emailSecondPresenter,
            xpFactor,
            description,
        } = body;

        // Check for unique title
        const existingSession = await this.sessionRepository.findOne({
            title,
        });
        if (existingSession) {
            throw new SessionTitleAlreadyInUse();
        }

        const session = new Session();
        session.title = title;
        session.subTitle = subTitle;
        session.xpFactor = xpFactor;
        session.description = description;

        if (emailFirstPresenter === emailSecondPresenter) {
            throw new SessionPresenterEmailsMustDiffer();
        }
        const presenterEmails = emailSecondPresenter
            ? [emailFirstPresenter, emailSecondPresenter]
            : [emailFirstPresenter];
        const presenters = await Promise.all(
            presenterEmails.map(email => this.getPossiblyExistingUser(email)),
        );

        session.firstPresenter = presenters[0];
        session.secondPresenter = presenters[1];
        await this.sessionRepository.save(session);

        this.sendSessionCreatedMails(body, presenters, origin);
    }

    async updateSession(
        body: UpdateSessionRequest,
        sessionId: string,
        user: IUserSession,
    ): Promise<string> {
        const {
            emailFirstPresenter,
            emailSecondPresenter,
            sessionState,
            personaIds,
        } = body;

        const existingSession = await this.sessionRepository.findOne({
            where: { id: sessionId },
            relations: [
                'firstPresenter',
                'secondPresenter',
                'intendedAudience',
            ],
        });
        if (!existingSession) {
            throw new SessionNotFound();
        }

        if (emailFirstPresenter === emailSecondPresenter) {
            throw new SessionPresenterEmailsMustDiffer();
        }
        if (
            emailFirstPresenter !== user.email &&
            emailSecondPresenter !== user.email &&
            !this.checkAdminRightsOnSession(user)
        ) {
            throw new SessionEditNotAllowed();
        }

        if (this.checkAdminRightsOnSession(user)) {
            existingSession.sessionState = sessionState || null;

            existingSession.firstPresenter = await this.userRepository.findOne({
                email: emailFirstPresenter,
            });

            existingSession.secondPresenter =
                (await this.userRepository.findOne({
                    email: emailSecondPresenter,
                })) || null;
        }

        const personas = personaIds
            ? await this.personaRepository.find({ id: In(personaIds) })
            : [];
        if (personas.length) existingSession.intendedAudience = personas;

        existingSession.updatedBy = user.email;

        await this.sessionRepository.save(
            this.getRemainingUpdateValues(existingSession, body),
        );

        return existingSession.id;
    }

    /**
     * Check if a user with the given email exists, if so, return that user.
     * If there is no user found for email, create a new one.
     * Does not save the user to the database.
     * @param email Email of the new presenter
     */
    private async getPossiblyExistingUser(email: string): Promise<User> {
        const existingUser = await this.userRepository.findOne({ email });

        if (existingUser) {
            return existingUser;
        }

        const newUser = new User();
        newUser.email = email;
        const defaultRole = await this.roleRepository.findOne({
            isDefault: true,
        });
        newUser.roles = [defaultRole];
        newUser.resetToken = await this.createResetTokenForPresenter(email);
        newUser.state = UserState.Registering;
        return newUser;
    }

    private async createResetTokenForPresenter(email: string): Promise<string> {
        const resetToken = await this.jwtService.signAsync(
            { email },
            { expiresIn: '1d' },
        );
        return resetToken;
    }

    private sendSessionCreatedMails(
        body: CreateSessionRequest,
        presenters: User[],
        origin: string,
    ): void {
        // Mail the presenter(s) to notify that he was assigned to a new session
        this.mailerService.sendMail(
            sessionCreatedForFirstPresenterMessage({
                email: body.emailFirstPresenter,
                sessionTitle: body.title,
                frontendUrl: origin,
            }),
        );
        if (body.emailSecondPresenter) {
            this.mailerService.sendMail(
                sessionCreatedForSecondPresenterMessage({
                    email: body.emailSecondPresenter,
                    sessionTitle: body.title,
                    frontendUrl: origin,
                    emailMainPresenter: body.emailFirstPresenter,
                }),
            );
        }

        // If the presenter is a new user, we want to send him a registration email
        presenters
            .filter(presenter => presenter.state === UserState.Registering)
            .forEach(presenter =>
                this.mailerService.sendMail(
                    registerMessage({
                        email: presenter.email,
                        resetToken: presenter.resetToken,
                        frontendUrl: origin,
                    }),
                ),
            );
    }

    // We want to check if the user has admin rights to edit certain values of a session
    private checkAdminRightsOnSession(user: IUserSession): boolean {
        return user.permissions.sessions.admin;
    }

    private getRemainingUpdateValues(
        existingSession: Session,
        request: UpdateSessionRequest,
    ): Session {
        const session: Session = existingSession;

        session.subTitle = request.subTitle || null;
        session.description = request.description;
        session.xpFactor = request.xpFactor || 0;
        session.shortDescription = request.shortDescription || null;
        session.goal = request.goal || null;
        session.type = request.type || null;
        session.topic = request.topic || null;
        session.maxParticipants = request.maxParticipants || 50;
        session.duration = request.duration || null;
        session.laptopsRequired = request.laptopsRequired;
        session.otherLimitations = request.otherLimitations || null;
        session.roomSetup = request.roomSetup || null;
        session.neededMaterials = request.neededMaterials || null;
        session.expierenceLevel = request.expierenceLevel || null;
        session.outline = request.outline || null;
        session.materialDescription = request.materialDescription || null;
        session.materialUrl = request.materialUrl || null;

        return session;
    }
}
