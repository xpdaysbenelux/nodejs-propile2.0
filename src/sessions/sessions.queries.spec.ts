import { Test } from '@nestjs/testing';
import * as faker from 'faker';

import { SessionRepository, DatabaseModule } from '../database';
import { SessionsQueries } from './sessions.queries';

describe('SessionsQueries', () => {
    let sessionRepository: SessionRepository;
    let sessionsQueries: SessionsQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [SessionsQueries],
        }).compile();

        sessionsQueries = module.get(SessionsQueries);
        sessionRepository = module.get(SessionRepository);
    });

    afterAll(() => {
        sessionRepository.manager.connection.close();
    });

    describe('getSession', () => {
        it('should return the requested session correctly', async () => {
            const result = await sessionsQueries.getSession(
                '8c87028d-31f6-414b-b02d-e7431b6a7243',
            );
            expect(result).toMatchSnapshot();
        });

        it('should return nothing if the requested session does not exist', async () => {
            const result = await sessionsQueries.getSession(
                faker.random.uuid(),
            );
            expect(result).toMatchSnapshot();
        });
    });

    describe('getSessions', () => {
        it('should return a paged list of sessions', async () => {
            const list = await sessionsQueries.getSessions();
            expect(list).toMatchSnapshot();
        });
    });
});
