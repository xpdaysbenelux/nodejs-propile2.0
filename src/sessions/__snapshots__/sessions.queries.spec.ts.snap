// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionsQueries getSession should return nothing if the requested session does not exist 1`] = `undefined`;

exports[`SessionsQueries getSession should return the requested session correctly 1`] = `
Session {
  "createdAt": 2020-01-01T08:21:59.730Z,
  "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt.",
  "duration": "30",
  "expierenceLevel": "NOVICE",
  "firstPresenter": User {
    "email": "user2@propile.com",
    "id": "45c2c554-b06a-4209-aba8-b097cb5f8566",
  },
  "goal": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
  "id": "8c87028d-31f6-414b-b02d-e7431b6a7243",
  "laptopsRequired": false,
  "materialDescription": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
  "materialUrl": "www.test_material_url.com",
  "maxParticipants": 25,
  "neededMaterials": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor.",
  "otherLimitations": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.",
  "outline": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu.",
  "roomSetup": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
  "secondPresenter": User {
    "email": "user1@propile.com",
    "id": "49affd08-f461-4e80-899f-7612e971ff12",
  },
  "shortDescription": "Short desc, leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet.",
  "subTitle": "My experience about...",
  "title": "How to...",
  "topic": "TEAM_AND_INDIVIDUAL",
  "type": "OTHER",
  "xpFactor": 2,
}
`;

exports[`SessionsQueries getSessions should return a paged list of sessions 1`] = `
Object {
  "data": Array [
    Session {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt.",
      "duration": "30",
      "expierenceLevel": "NOVICE",
      "firstPresenter": User {
        "email": "user2@propile.com",
        "id": "45c2c554-b06a-4209-aba8-b097cb5f8566",
      },
      "goal": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
      "id": "8c87028d-31f6-414b-b02d-e7431b6a7243",
      "laptopsRequired": false,
      "materialDescription": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
      "materialUrl": "www.test_material_url.com",
      "maxParticipants": 25,
      "neededMaterials": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor.",
      "otherLimitations": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.",
      "outline": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu.",
      "roomSetup": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
      "secondPresenter": User {
        "email": "user1@propile.com",
        "id": "49affd08-f461-4e80-899f-7612e971ff12",
      },
      "shortDescription": "Short desc, leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet.",
      "subTitle": "My experience about...",
      "title": "How to...",
      "topic": "TEAM_AND_INDIVIDUAL",
      "type": "OTHER",
      "xpFactor": 2,
    },
    Session {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
      "duration": null,
      "expierenceLevel": "NOVICE",
      "firstPresenter": User {
        "email": "user2@propile.com",
        "id": "45c2c554-b06a-4209-aba8-b097cb5f8566",
      },
      "goal": null,
      "id": "05909211-9301-4e65-8b1e-65c00b672ee8",
      "laptopsRequired": false,
      "materialDescription": "",
      "materialUrl": "",
      "maxParticipants": 50,
      "neededMaterials": "",
      "otherLimitations": "",
      "outline": "",
      "roomSetup": "",
      "secondPresenter": null,
      "shortDescription": null,
      "subTitle": "How I...",
      "title": "Avoid the Top 10 mistakes made by...",
      "topic": "CASES_AND_INTROS",
      "type": "HANDS_ON",
      "xpFactor": 5,
    },
    Session {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
      "duration": null,
      "expierenceLevel": "NOVICE",
      "firstPresenter": User {
        "email": "development@icapps.com",
        "id": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
      },
      "goal": null,
      "id": "9a72df45-46cc-45f3-ad63-548df4a385c5",
      "laptopsRequired": false,
      "materialDescription": null,
      "materialUrl": null,
      "maxParticipants": 50,
      "neededMaterials": null,
      "otherLimitations": null,
      "outline": null,
      "roomSetup": null,
      "secondPresenter": null,
      "shortDescription": null,
      "subTitle": null,
      "title": "Don't Be Fooled By...",
      "topic": null,
      "type": null,
      "xpFactor": 0,
    },
  ],
  "meta": Object {
    "count": 3,
    "skip": null,
    "totalCount": 3,
  },
}
`;
